drivers/gpio/gpio_pcf8574.c

```diff
@@ -0,0 +1,397 @@
+		LOG_ERR("Pin(s) is/are configured as output which should be input.");
+		return -EOPNOTSUPP;
+	}
+
+	k_sem_take(&drv_data->lock, K_FOREVER);
+
+	/**
+	 * Reading of the input port also clears the generated interrupt,
+	 * thus the configured callbacks must be fired also here if needed.
+	 */
+	rc = pcf8574_process_input(dev, value);
+
+	k_sem_give(&drv_data->lock);
+
+	return rc;
+}
+
+/**
+ * @brief This function realizes the write connection to the i2c device.
+ *
+ * @param dev A pointer to the device structure
+ * @param mask A mask of bits to set some bits to LOW or HIGH
+ * @param value The value which is written via i2c to the pfc8574's output pins
+ * @param toggle A way to toggle some bits with xor
+ *
+ * @retval 0 If successful.
+ * @retval Negative value for error code.
+ */
+static int pcf8574_port_set_raw(const struct device *dev, uint8_t mask, uint8_t value,
+				uint8_t toggle)
+{
+	const struct pcf8574_drv_cfg *drv_cfg = dev->config;
+	struct pcf8574_drv_data *drv_data = dev->data;
+	int rc = 0;
+	uint8_t tx_buf;
+
+	/** Make sure, kernel isn't busy*/
+	if (k_is_in_isr()) {
+		return -EWOULDBLOCK;
+	}
+
+	if ((drv_data->pins_cfg.configured_as_outputs & value) != value) {
+		LOG_ERR("Pin(s) is/are configured as input which should be output.");
+		return -EOPNOTSUPP;
+	}
+
+	tx_buf = (drv_data->pins_cfg.outputs_state & ~mask);
+	tx_buf |= (value & mask);
+	tx_buf ^= toggle;
+
+	rc = i2c_write_dt(&drv_cfg->i2c, &tx_buf, sizeof(tx_buf));
+
+	if (rc != 0) {
+		LOG_ERR("%s: failed to write output port: %d", dev->name, rc);
+		return -EIO;
+	}
+
+	k_sem_take(&drv_data->lock, K_FOREVER);
+	drv_data->pins_cfg.outputs_state = tx_buf;
+	k_sem_give(&drv_data->lock);
+
+	return 0;
+}
+
+/**
+ * @brief This function fills a dummy because the pfc8574 has no pins to configure.
+ * You can use it to set some pins permanent to HIGH or LOW until reset. It uses the port_set_raw
+ * function to set the pins of pcf8574 directly.
+ *
+ * @param dev Pointer to the device structure for the driver instance.
+ * @param pin The bit in the io register which is set to high
+ * @param flags Flags like the GPIO direction or the state
+ *
+ * @retval 0 If successful.
+ * @retval Negative value for error.
+ */
+static int pcf8574_pin_configure(const struct device *dev, gpio_pin_t pin, gpio_flags_t flags)
+{
+	struct pcf8574_drv_data *drv_data = dev->data;
+	int ret = 0;
+	uint8_t temp_pins = drv_data->pins_cfg.outputs_state;
+
+	if (flags & (GPIO_PULL_UP | GPIO_PULL_DOWN | GPIO_DISCONNECTED | GPIO_SINGLE_ENDED)) {
+		return -ENOTSUP;
+	}
+	if (flags & GPIO_OUTPUT_INIT_HIGH) {
+		temp_pins |= (1 << pin);
+	}
+	if (flags & GPIO_OUTPUT_INIT_LOW) {
+
+		temp_pins &= ~(1 << pin);
+	}
+	if (flags & GPIO_OUTPUT) {
+		drv_data->pins_cfg.configured_as_outputs |= BIT(pin);
+	}
+	if (flags & GPIO_INPUT) {
+		drv_data->pins_cfg.configured_as_outputs &= ~BIT(pin);
+	}
+
+	ret = pcf8574_port_set_raw(dev, temp_pins, temp_pins, 0);
```

Comment:
I think the mask still needs to be applied?

```suggestion
	if (flags & GPIO_INPUT) {
		drv_data->pins_cfg.configured_as_outputs &= ~BIT(pin);
	} else if (flags & GPIO_OUTPUT) {
		if (flags & GPIO_OUTPUT_INIT_HIGH) {
			temp_pins |= (1 << pin);
		}
		if (flags & GPIO_OUTPUT_INIT_LOW) {
			temp_pins &= ~(1 << pin);
		}
		drv_data->pins_cfg.configured_as_outputs |= BIT(pin);
	}

	ret = pcf8574_port_set_raw(dev,
				   drv_data->pins_cfg.configured_as_outputs,
				   temp_pins,
				   0);
```

Reply:
I think the if statements in else if could be reached only if flags is set to  GPIO_OUTPUT and GPIO_OUTPUT_INIT_HIGH or LOW in the same call, but a user might set it in 2 different calls.